### PR TITLE
[DS-4153] Added repository createAndReturn that accepts parent id

### PR DESCRIPTION
**Pull request by mspalti**
https://jira.duraspace.org/browse/DS-4153
I decided to add a createAndReturn method that takes the parent id as a parameter since I didn't see a better approach. The parent id is retrieved in the controller.  I'm using the id to check permissions on the parent object when adding collections and sub-communities.

### CODE REVIEW

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CollectionRestRepository.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CollectionRestRepository.java]: While this error message is accurate, it looks odd in this class (as there's no `parent` param anywhere in the `CollectionRestRepository` anymore.  So, I think we should reword this slightly to say something like:
`"Cannot create a Collection without providing a parent Community"`

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CollectionRestRepository.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CollectionRestRepository.java]: If the `id` is null here, we should return a message that says it's null (i.e. there's no point in printing out the value (as it'll always be null).  So, maybe reword this error to be similar to the one above (`"Parent Community UUID is null. Cannot create a Collection without providing a parent Community"`)

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java]: Similar to above, the `id` is known to be null.  Maybe reword this to be something like `"Parent Community UUID is null. Cannot create a SubCommunity without providing a parent Community"`

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java]: This is accidentally swallowing part of the exception and only returning the main error message (which is all the `toString()` will provide).  Sometimes that's "good enough", but other times, more info can be found from the entire stack trace.  I prefer to return the entire stacktrace like this:
`throw new UnprocessableEntityException("Error parsing request body", e1);`

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java]: I noticed we are missing new ITs to prove the new Collection createAndReturn is checking permissions properly.  They should be similar to the Community ITs you've added, but we should have ITs for both objects.

**Review by tdonohue — COMMENTED** [review]
@mspalti : Overall this code looks great to me!  I only have some minor requests to enhance/improve some of the error messages you are returning & add a few more ITs (in `CollectionRestRepositoryIT`). I like the implementation here though, and I appreciate the new Community tests to prove it works

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java
Comment (mspalti) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java]: Oops.  That's a big omission.  Tests added.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java
Comment (mspalti) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunityRestRepository.java]: Done.  Updated the exception in CollectionRestRepository to also return the stack trace.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CollectionRestRepositoryIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CollectionRestRepositoryIT.java]: After this line we should have a `context.restoreAuthSystemState();` to turn the authorization system back "on". It only needs to be turned off temporarily to create the test data...after that, we need to be sure the tests themselves are obeying all authorization rules.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CollectionRestRepositoryIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CollectionRestRepositoryIT.java]: Again, after this line we should have a `context.restoreAuthSystemState();` to turn the authorization system back "on".

**Review by tdonohue — COMMENTED** [review]
Looking good overall @mspalti!  However, I noticed a few other minor issues in the new CollectionRestRepositoryIT tests.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CollectionRestRepositoryIT.java
Comment (mspalti) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/CollectionRestRepositoryIT.java]: @tdonohue ,  fixing these test problems in the next commit.

Looking things over, I notice quite a few tests to not restoreAuthSystemState, which surprised me.  They're not all simple GET requests, and some go to the trouble of passing a token.   So I'm a little confused about the effect of restoring authorization in the integration tests. It seems like it should always be done (even if not testing authorization) but it's not...

Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/CollectionRestRepositoryIT.java]: @mspalti : Yes, restoring authorization should always be done. I try to look for it in reviews, but may not always catch it.  While in most scenarios the restoring of authorization may not be a big deal either way, it could turn out to be problematic to maintenance if later code is added that is affected by the lack of authorization checks -- i.e. it's very possible a test could be succeeding simply cause authorization checks are disabled, and *not* because it actually "works".

In any case, feel free to clean up others if you ever find them around.  I do try to check for it in all code reviews, but some may have slipped through the cracks.

Comment (mspalti) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/CollectionRestRepositoryIT.java]: I updated the community and collection repository tests by restoring authorization where i found it missing.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java
Comment (cwilper) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java]: postJsonInternal is a generic method. The dso type is not known at this point. It *may* have a parent. I would therefore call this parentObjectId or parentObjectUuid.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java
Comment (cwilper) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java]: What about just having `parentUuid` be an arg to this method? Then caller (post(..)) could pass it in, having gotten it from a `@RequestParam` argument (with optional = true). It would make the contract here "You can optionally pass in a parent dso uuid") more explicit without having to read the body of the code.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityRestRepositoryIT.java
Comment (cwilper) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityRestRepositoryIT.java]: 👍 This is a good test to see, as it shows we can indeed do deeper authorization checks than "user is admin" with the annotations.

**Review by cwilper — APPROVED** [review]
Approach looks good and test coverage looks good. I just had a couple minor comments about args and naming but generally approve of this.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @mspalti for the hard work on this.  This looks good to me now.  However, I do agree with @cwilper 's minor changes noted above.  Did you want to perhaps quickly make those changes or otherwise respond to them before we merge this?

**Comment by mspalti** [comment]
Sorry for the delay, @tdonohue .  I added the suggestions by @cwilper to the PR. Should be ready to merge after tests.

**Comment by tdonohue** [comment]
Thanks @mspalti for the quick updates based on @cwilper's feedback.  I like these minor refactors (and they look to be exactly what Chris asked for).  As this is at +2, merging immediately.  Thanks again for the hard work on this!

**Pull request merged**
